crates/app/src/app.rs

```diff
@@ -0,0 +1,225 @@
+use std::path::PathBuf;
+
+use serde::Deserialize;
+use spin_core::{Engine, StoreBuilder};
+use url::Url;
+
+use crate::locked::{
+    ContentPath, LockedApp, LockedComponent, LockedComponentSource, LockedTrigger,
+};
+use crate::values::MetadataExt;
+use crate::{AppLoader, Error, Result, Variable};
+
+/// An `App` holds loaded configuration for a Spin application.
+#[derive(Debug)]
+pub struct App<'a> {
+    pub(crate) loader: &'a AppLoader,
+    pub(crate) uri: String,
+    pub(crate) locked: LockedApp,
+}
+
+impl<'a> App<'a> {
+    /// Returns a [`Loader`]-implementation-specific URI for this app.
+    pub fn uri(&self) -> &str {
+        &self.uri
+    }
+
+    /// Deserializes typed metadata for this app.
+    ///
+    /// Returns `Ok(None)` if there is no metadata for the given `key` and an
+    /// `Err` only if there _is_ a value for the `key` but the typed
+    /// deserialization failed.
+    pub fn get_metadata<'this, T: Deserialize<'this>>(&'this self, key: &str) -> Result<Option<T>> {
+        self.locked.metadata.get_typed(key)
+    }
+
+    /// Deserializes typed metadata for this app.
+    ///
+    /// Like [`App::get_metadata`], but returns an error if there is
+    /// no metadata for the given `key`.
+    pub fn require_metadata<'this, T: Deserialize<'this>>(&'this self, key: &str) -> Result<T> {
+        self.locked.metadata.require_typed(key)
+    }
+
+    /// Returns the path to the origin manifest file, if present.
+    pub fn file_origin(&self) -> Result<Option<PathBuf>> {
+        match self.get_metadata::<Url>("origin")? {
+            Some(url) if url.scheme() == "file" => {
+                let path = url.to_file_path().map_err(|_| {
+                    Error::MetadataError(format!("Invalid origin file URL: {url:?}"))
+                })?;
+                Ok(Some(path))
+            }
+            _ => Ok(None),
+        }
+    }
```

Comment:
This file is mostly just code moved from `lib.rs`, but I also added this method.

Reply:
...which I just realized I don't actually use in this revision :facepalm: